internal/index/indexscanner/scanner.go

```diff
@@ -98,26 +98,45 @@ func ReadPluginFromFile(path string) (index.Plugin, error) {
 
 func ReadPlugin(f io.ReadCloser) (index.Plugin, error) {
 	defer f.Close()
-	p, err := DecodePluginFile(f)
+	plugin := &index.Plugin{}
+	err := decodeFile(f, plugin)
+	p := *plugin
 	if err != nil {
 		return p, errors.Wrap(err, "failed to decode plugin manifest")
 	}
 	return p, errors.Wrap(validation.ValidatePlugin(p.Name, p), "plugin manifest validation error")
 }
 
-// DecodePluginFile tries to decodes a plugin manifest from r.
-func DecodePluginFile(r io.Reader) (index.Plugin, error) {
-	var plugin index.Plugin
+// ReadReceiptFromFile loads a file from the FS. When receipt file not found, it
+// returns an error that can be checked with os.IsNotExist.
+func ReadReceiptFromFile(path string) (index.Receipt, error) {
+	f, err := os.Open(path)
+	if os.IsNotExist(err) {
+		// TODO(ahmetb): we should use go1.13+ errors.Is construct at call sites to evaluate if an error is os.IsNotExist
+		return index.Receipt{}, err
+	} else if err != nil {
+		return index.Receipt{}, errors.Wrap(err, "failed to open index file")
+	}
```

Comment:
as far as I can tell this can still be deduplicated (with the same de-dupe technique as `decodeFile` you've done).